src/mysky/encrypted_files.ts

```diff
@@ -0,0 +1,330 @@
+import randomBytes from "randombytes";
+import { sanitizePath } from "skynet-mysky-utils";
+import { secretbox } from "tweetnacl";
+
+import { HASH_LENGTH, sha512 } from "../crypto";
+import { JsonData, JsonFullData } from "../skydb";
+import { hexToUint8Array, stringToUint8ArrayUtf8, toHexString, uint8ArrayToStringUtf8 } from "../utils/string";
+import {
+  throwValidationError,
+  validateBoolean,
+  validateHexString,
+  validateNumber,
+  validateObject,
+  validateString,
+  validateUint8Array,
+  validateUint8ArrayLen,
+} from "../utils/validation";
+
+export const ENCRYPTED_JSON_RESPONSE_VERSION = 1;
+
+/**
+ * The length of the encryption key.
+ */
+export const ENCRYPTION_KEY_LENGTH = 32;
+
+/**
+ * The length of the hidden field metadata stored along with encrypted files.
+ * Note that this is hidden from the user, but not actually encrypted. It is
+ * stored after the nonce.
+ */
+export const ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH = 16;
+
+/**
+ * The length of the random nonce, prepended to the encrypted bytes. It comes
+ * before the unencrypted hidden field metadata.
+ */
+export const ENCRYPTION_NONCE_LENGTH = 24;
+
+/**
+ * The length of the overhead introduced by encryption.
+ */
+const ENCRYPTION_OVERHEAD_LENGTH = 16;
+
+/**
+ * The length of the share-able path seed.
+ */
+export const ENCRYPTION_PATH_SEED_LENGTH = 32;
+
+export type EncryptedJSONResponse = {
+  data: JsonData | null;
+};
+
+type DerivationPathObject = {
+  pathSeed: Uint8Array;
+  directory: boolean;
+  name: string;
+};
+
+type EncryptedFileMetadata = {
+  // 8-bit uint.
+  version: number;
+};
+
+/**
+ * Decrypts the given bytes as an encrypted JSON file.
+ *
+ * @param data - The given raw bytes.
+ * @param key - The encryption key.
+ * @returns - The JSON data and metadata.
+ * @throws - Will throw if the bytes could not be decrypted.
+ */
+export function decryptJSONFile(data: Uint8Array, key: Uint8Array): JsonFullData {
+  validateUint8Array("data", data, "parameter");
+  validateUint8ArrayLen("key", key, "parameter", ENCRYPTION_KEY_LENGTH);
+
+  // Validate that the size of the data corresponds to a padded block.
+  if (!checkPaddedBlock(data.length)) {
+    throwValidationError("data", data, "parameter", `padded encrypted data, length was '${data.length}'`);
+  }
+
+  // Extract the nonce.
+  const nonce = data.slice(0, ENCRYPTION_NONCE_LENGTH);
+  data = data.slice(ENCRYPTION_NONCE_LENGTH);
```

Comment:
I tried but it made the code more verbose without really reducing the potential for mistakes. In fact when I had multiple variables like `data`, `dataWithoutNonce`, `dataWithoutMetadata` it just made it more confusing.

Reply:
```js
const nonce = data.slice(0, ENCRYPTION_NONCE_LENGTH);
const metadataBytes = data.slice(ENCRYPTION_NONCE_LENGTH, ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
const dataBytes = data.slice(ENCRYPTION_NONCE_LENGTH + ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
```

that would probably work and you wouldn't have to overwrite function argument
ready why here https://eslint.org/docs/rules/no-param-reassign

Reply:
That's fair enough. I can't enable that lint here because it would cause a lot of warnings and errors unrelated to this PR. I'll write a follow-up. This is a style issue and doesn't block this PR.

Reply:
https://github.com/SkynetLabs/skynet-js/issues/186